docs/design/02-topologies.md

```diff
@@ -0,0 +1,288 @@
+all the defined groups, along with the nodes within each group and the config
+for each node (to get each node's IP and hostname etc).
+
+A possible way to do this would be to have an `alces.groups` method, which
+takes a block with 2 arguments, `group_index` and `group`. `group_index` will
+simply be a unique index for each configured group; `group` will have a `nodes`
+method which takes a block with a `node` argument; `node` will provide access
+to the loaded config for the given node, the same as would be available at the
+top level if the templater was run with the `nodename` for this node and an
+appropriate `index`.
+
+This could then be used to generate a `hosts` file in some way, possibly like
+this:
+
+```erb
+<%# Note: may not actually need to provide group_index here, just make it
+available in configs. %>
+<% alces.groups do |group_index, group| %>
+
+  <% group.nodes do |node| %>
+
+    <% node.networks.each do |name, network| %>
+      <%# Note: `network.ip` could be either hard-coded or generated
+      dynamically using `group_index` and `index` etc, will depend on config.
+      %>
+      <%= network.ip %> <%= network.hostname %> <%= network.short_hostname %>
+    <% end %>
+
+  <% end %>
+
+<% end %>
+```
+
+The `alces.groups` parameter in this situation is going to load the config for
+every possible node in every configured group for the Cluster. There is the
+potential that this will be noticeably slow; if this is the case we may want to
+have this only available in certain situations, for example when a
+`load_groups`, or similar, argument is passed to the templater.
+
+When this feature is implemented the `hosts` file will be able to be rendered
+in full in a single pass; since the information used in the rendering can
+change with any `configure` command it seems sensible to automatically
+re-render the `hosts` file after every `configure` command, similarly to the
+plan for `genders` after any `configure group` command. If we add this it will
+largely make the `hosts` command obsolete, although we may want to retain it
+for re-rendering after manually changing the template or a config file.
+
+We will also need to consider if we this how a `hosts` template other than the
+default will be selected for automatic re-rendering, or alternatively if we
+still need the ability to have multiple `hosts` templates within a single repo
+now - it should now be possible to have the `hosts` template handle any
+situations where we want to treat groups or nodes differently.
+
+
+## Command ordering
+
+With this new functionality certain commands will now require that other
+commands have been run first:
+
+- All `metal configure` commands will require the repo be set up (`repo use`
+  has been run) before running.
+
+- `metal build` will require that `metal configure all` has been run first,
+  i.e. that `/var/lib/metalware/config/all.yaml` has been generated, so any
+  necessary config has been set up before templating.
```

Comment:
It might be hard to define it in the cli yaml config. Think of it as how `Commander` behaves and not how `BaseCommand` works. It's probably just easier to define a `DEPENDENCIES_LIST` in the inherited classes. Then in the `BaseCommand` check if the `DEP` constant exists. If it doen't then their is no dependencies. If it does then enforce them.